### DISCUSSION TITLE
Any serious competitors?

### DISCUSSION THREAD

**Discussion by jasonkuhrt** [grafana/tempo#3639] [General]
Hey, I'm evaluating Tempo for its TraceQL features and how that translates into metrics dashboards etc.

In looking at the market, best I can tell, TraceQL is far and away the most sophisticated querying language for traces. Can people more familiar with this space confirm this? I would also appreciate hearing what people consider to be the closest second, third, ... technology alternatives here. Thanks!

**Comment by joe-elliott** [comment]
I have my eye on this discussion and am hoping someone else chimes in. I can't say I'm the most neutral source to answer if there are "any serious competitors" to TraceQL :)

FWIW, TraceQL was designed from the ground up as a tracing first query language. In particular we highly valued the ability to concisely express structural conditions in the query. Other vendors and OSS projects attempt this with "SQL like" languages but I personally have always disliked how SQL maps into this domain. Hence, TraceQL.

I'll give space for others to give answers.

**Reply by jasonkuhrt** [reply]
From what I have seen, other vendors are also trying to re-purpose their query languages for logs for tracing, then spinning it as a "unified system". I have yet to see a pitch wherein the vendor clearly articulates how their so-called unification is benefiting operations around tracing data. It appears to be a marketing pitch for people working on systems with "traditional" instrumentation.

I would be very curious to see someone point out a vendor that documents how their system really integrates with the incoming telemetry data structures of OTEL (tracing, and for that matter metrics and events too). So far only Grafana Tempo with TraceQL has paid attention here on the OTEL trace aspect from the ground up. Am I wrong?

**Reply by joe-elliott** [reply]
You have hit on another important aspect of TraceQL. It was definitely designed around the OpenTelemetry object model. Those familiar OTel will spin up more quickly on the language.

Tempo was likely the first OTel native database. From the very beginning we supported native gRPC OTel ingestion.

And I think TraceQL is a candidate for an OTel native query language, but I'm not sure that concept exists or OTel would agree :).

**Reply by jasonkuhrt** [reply]
@joe-elliott is there any vision where TraceQL is OSS'd such that it could be neutral, have a spec, be adopted by multiple vendors? Or will it remain a Grafana proprietary technology?

**Reply by joe-elliott** [reply]
I have tried to reach out to several groups about query standardization but there doesn't seem to be much appetite for it in the industry. OTel is not interested in defining a query language. There is a [CNCF Tag](https://github.com/cncf/tag-observability/tree/main/working-groups/query-standardization) for query standardization but it doesn't have momentum.

If there were a serious effort at standardization I would push for TraceQL to be in consideration.
